### PR TITLE
Linear advance with I2S streaming

### PR DESCRIPTION
**Pull request by tombrazier**
### Description

This PR does the following:
- Add option `EXPERIMENTAL_I2S_LA` to experimentally override the sanity check which prevents `I2S_STEPPER_STREAM` from compiling with `LIN_ADVANCE`.
- Modifies linear advance and the I2S code to make them work together.

It also removes an off-by-one error in the number of ticks between stepper pulses. (The variable storing time remaining previously was only decremented for ticks where the stepper pin was low, missing the one tick in each cycle where the pin went high.)

### Requirements

ESP32 boards.

### Benefits

Adds Linear Advance to ESP32 boards.

### Configurations

[config.zip](https://github.com/MarlinFirmware/Marlin/files/9427539/config.zip)

### CODE REVIEW

**Review by tombrazier — COMMENTED** [review]
[Edit: this comment was supposed to be a review of 7ab0500 - somehow it ended up as a comment instead.]

`i2s_push_sample()` has to be called every time, regardless of whether a step was generated. `stepperTask()` in `Marlin/src/HAL/ESP32/i2s.cpp` expects it.

File: Marlin/src/HAL/ESP32/i2s.cpp [hunk 1]
```diff
@@ -139,21 +139,44 @@ static void IRAM_ATTR i2s_intr_handler_default(void *arg) {
...
+      if (!nextMainISR) {
         Stepper::pulse_phase_isr();
-        remaining = Stepper::block_phase_isr();
+        nextMainISR = Stepper::block_phase_isr();
+      }
+
+      LOOP_L_N(i, push_count) {
+        nextMainISR--;
+        #if ENABLED(LIN_ADVANCE)
+          if (nextAdvanceISR != Stepper::LA_ADV_NEVER)
```
Comment (thinkyhead) [inline Marlin/src/HAL/ESP32/i2s.cpp]: Are you counting on `nextAdvanceISR` being equal to -1 here, as a result of the decrement, and for `LA_ADV_NEVER` to also be equal to -1 as part of this logic? I'm going to try and just use subtraction here without the loop, even though the compiler might be unrolling.

Comment (tombrazier) [reply Marlin/src/HAL/ESP32/i2s.cpp]: No I am not. I don't expect the decrement to go below zero.

**Comment by tombrazier** [comment]
> [don't use TERN_ for declarations](https://github.com/MarlinFirmware/Marlin/pull/24684/commits/f6ba3655811052d511c2d8100504f5e6e5a5e86b)

Will try to remember that.

**Comment by thinkyhead** [comment]
> Will try to remember that.

Yeah, it seems to mess with some editors' syntax parsers.

**Comment by EvilGremlin** [comment]
It's harder to read even with proper highlighting, too much punctuation for one line

**Comment by thinkyhead** [comment]
> `i2s_push_sample()` has to be called every time, regardless of whether a step was generated. `stepperTask()` in `Marlin/src/HAL/ESP32/i2s.cpp` expects it.

Ah, okay, I clearly brainfarted on that one.

**Comment by thinkyhead** [comment]
Ok, I'm done obsessing over every cycle and everything looks good. So… back to obsessing over input shaping, I guess!

**Comment by HoverClub** [comment]
Just looked at this and I -think- there is a possible issue with `push_sample()` being called without a check for buffer space the 2nd time?  Every iteration of `stepperTask()` (with one sample space in the buffer) should only call ` push_sample()` once.  If  `Stepper::pulse_phase_isr()` and `Stepper::advance_isr()` time out at same time then they'd need prioritising (the deferred one would get handled next `stepperTask()` loop when buffer space - 4us max).  Would also avoid the `push_count` stuff?

**Comment by tombrazier** [comment]
Without looking ito it in too much detail, that makes sense to me. Maybe `while (dma.rw_pos < DMA_SAMPLE_COUNT)` should be `while (dma.rw_pos < DMA_SAMPLE_COUNT - 1)`. I'm just guessing here. It's been a while since I looked at that code.

**Comment by HoverClub** [comment]
Yep - that should work.  It also fixes the -2 subtracts on the **next...ISRs** if they are value 1.

This should also work (and is simpler) but prioritises the stepper ISR over LA which could add a 4usec delay to the LA step pulses (that happens anyway with the existing code).
```C++
    while (dma.rw_pos < DMA_SAMPLE_COUNT) {
      // Fill with the port data post pulse_phase until the next step
      if (nextMainISR && TERN1(LIN_ADVANCE, nextAdvanceISR)) {
        i2s_push_sample();
        nextMainISR--;
        TERN_(LIN_ADVANCE, nextAdvanceISR--);
      }
      else if (!nextMainISR) {
        Stepper::pulse_phase_isr();
        nextMainISR = Stepper::block_phase_isr();
      }
      #if ENABLED(LIN_ADVANCE)
        else {
          if (!nextAdvanceISR) {
            Stepper::advance_isr();
            nextAdvanceISR = Stepper::la_interval;
          }
          else if (nextAdvanceISR == Stepper::LA_ADV_NEVER)
            nextAdvanceISR = Stepper::la_interval;
        }
      #endif
    }
```

**Comment by tombrazier** [comment]
That approach means that `nextAdvanceISR` doesn't get decremented whenever `Stepper::pulse_phase_isr()` is called and `nextMainISR` doesn't get decremented whenever `Stepper::advance_isr()` is called (which is a lot during accel/decel ramps). I think this might mess with step rates quite a bit.
Also
```
          else if (nextAdvanceISR == Stepper::LA_ADV_NEVER)
            nextAdvanceISR = Stepper::la_interval;
```
should happen every time through the loop unless it is redundant because
```
        if (!nextAdvanceISR) {
          Stepper::advance_isr();
          nextAdvanceISR = Stepper::la_interval;
        }
```
just happened.

**Comment by HoverClub** [comment]
This should work (again!).  I assume the test `(nextAdvanceISR == Stepper::LA_ADV_NEVER)` detects a dec below zero to trigger a reload of the starting value again?  In which case this code should still work even if the reloaded value == `Stepper::LA_ADV_NEVERStepper::LA_ADV_NEVER`.  If not, then the buffer-2 test is probably the easiest approach.
```C++
    while (dma.rw_pos < DMA_SAMPLE_COUNT) {

      if (!nextMainISR) {
        Stepper::pulse_phase_isr();
        nextMainISR = Stepper::block_phase_isr();
      }
      #if ENABLED(LIN_ADVANCE)
        else if (!nextAdvanceISR) {
          Stepper::advance_isr();
          nextAdvanceISR = Stepper::la_interval;
        }
      #endif
      else 
        i2s_push_sample();

      nextMainISR--;

      #if ENABLED(LIN_ADVANCE)
        if (nextAdvanceISR == Stepper::LA_ADV_NEVER)
          nextAdvanceISR = Stepper::la_interval;
        nextAdvanceISR--;
      #endif
    }

```

**Comment by tombrazier** [comment]
`Stepper::LA_ADV_NEVER` is a special value that indicates that the LA ISR is not required to be called at present. But if `Stepper::la_interval` gets set to a different value then `nextAdvanceISR` needs updating to this value. I believe the code above will work from this perspective. It is not intended that `nextAdvanceISR` ever decrements below zero and overflows to the value `Stepper::LA_ADV_NEVER`.

I think it does change timings, making them more accurate but also introducing the potential that the STEP pulse never goes low again. Previously, `nextAdvanceISR` and `nextMainISR` did not get decremented immediately after `Stepper::advance_isr();` or `Stepper::pulse_phase_isr();`, which meant that there would actually be an extra 4us period between pulses. This ensured there was space for the STEP pin to go low. I think you have already noticed this and are happy with the trade-off.

There is one actual real problem, though: if both trigger at the same time, then `nextAdvanceISR` will be decremented to -1, which, as it happens, is also the value of `Stepper::LA_ADV_NEVER`. That means in the next run through the loop, `Stepper::advance_isr();` won't get called.

If the buffer-2 option is used, will there be an issue if the last entry in the buffer does not get filled? I suspect that there would need to be a special `i2s_push_sample()` to make sure the whole buffer is populated.

**Comment by HoverClub** [comment]
It's quite a tricky bit of code!   This will fix the decrement-past-zero problem and another issue where `nextAdvancedISR` could be decremented after it has been re-loaded with `Stepper::LA_ADV_NEVER`  - resulting in the  `Stepper::advance_isr()` being called again when it should be disabled.


```C++    while (dma.rw_pos < DMA_SAMPLE_COUNT) {

      if (!nextMainISR) {
        Stepper::pulse_phase_isr();
        nextMainISR = Stepper::block_phase_isr();
      }
      #if ENABLED(LIN_ADVANCE)
        else if (!nextAdvanceISR) {
          Stepper::advance_isr();
          nextAdvanceISR = Stepper::la_interval;
        }
      #endif
      else 
        i2s_push_sample();

      nextMainISR--;

      #if ENABLED(LIN_ADVANCE)
        if (nextAdvanceISR == Stepper::LA_ADV_NEVER) 
          nextAdvanceISR = Stepper::la_interval;
    
        if (nextAdvanceISR && nextAdvanceISR != Stepper::LA_ADV_NEVER) 
          nextAdvanceISR--;
      #endif
    }
```

**Comment by tombrazier** [comment]
I think that's a pretty good compromise. If both fire at the same time, LA will lose one tick but that's not a very serious problem and it will be corrected over time by logic in stepper.cpp.

**Comment by HoverClub** [comment]
Good.  The missing LA tick is just a delay - it'll get caught on next loop iteration provided the system isn't stepping at max rate (125KHz?).  All it'll do is add 4us jitter to the LA stepper - nothing compared to the hardware latency.

I'll leave the Pull req to you if you don't mind - I've a habit of screwing up GitHub stuff!  

I've tested this in my system and there are no discernable issues as you'd expect.

**Comment by nikybiasion** [comment]
Hi, is this pull request a workaround to use linear advance with boards like MKS tinybee? I've tried the bugfix 2.1.x but when i print the linear advance calibration test, only the first line with K=0 is printed, in the other lines the motor extruder not spin.
I've tried to replace the last code in the third last comment, first line was printed, in the second line the extruder works, but the print of the line does not stop, X motor still spin to the bed endstop, i need to turn off the printer.
I use TMC2209

**Comment by tombrazier** [comment]
> Hi, is this pull request a workaround to use linear advance with boards like MKS tinybee?

Yes it is. It really needs someone with an ESP32 to follow up.

**Comment by tombrazier** [comment]
> I'll leave the Pull req to you if you don't mind - I've a habit of screwing up GitHub stuff!

@HoverClub I don't actually have an ESP32 board and I can't remember how I ended up writing this PR. I also have a bunch of other TODOs. I think it needs someone with the actual hardware to take over...

**Comment by HoverClub** [comment]
> I don't actually have an ESP32 board and I can't remember how I ended up writing this PR. I also have a bunch of other TODOs. I think it needs someone with the actual hardware to take over...

I use a TinyBee ESP32 printer so can re-test it and issue the pull - may take a week or two atm though!

**Comment by tombrazier** [comment]
Go for it. You'll get there a long time before me.

**Comment by nikybiasion** [comment]
I've tried EXPERIMENTAL_I2S_LA with MKS Tinybee and TMC2209, when K differ from 0, the print start, but when the extruder start moving along X axis, it does not stop and continue over the bed limit, and i have to stop the print...
This, with #25113 pull request
Without #25113, extruder not extrude, but the movements are right

**Comment by HoverClub** [comment]
> I've tried EXPERIMENTAL_I2S_LA with MKS Tinybee and TMC2209, when K differ from 0, the print start, but when the extruder start moving along X axis, it does not stop and continue over the bed limit, and i have to stop the print... This, with #25113 pull request Without #25113, extruder not extrude, but the movements are right

Maybe best for you to start a new issue with this so your setup detail can be seen?  I have a similar setup (TinyBee + TMC2209) and it all works as it should - except input shaping which isn't compatible atm.

**Pull request merged**